### PR TITLE
server: support custom URI schemes in CORS config (fix #5818)

### LINKED ISSUE #5818: server: Support custom URI schemes in CORS config
It would be great if we could specify custom URI schemes in the `HASURA_GRAPHQL_CORS_DOMAIN` list. Right now it only supports `http://` and `https://`.

We have recently started using Ionic's [CapacitorJS](https://capacitorjs.com/) to build our webapp into mobile apps (inside a webview). The following scheme is used for iOS: `capacitor://localhost`,  but I am unable to add it to Hasura's CORS config.

See following error:
```
Fatal Error:- Environment variable HASURA_GRAPHQL_CORS_DOMAIN: invalid domain: 'capacitor://localhost'. All domains should have scheme + (optional wildcard) host + (optional port)
```

I also tried to use a wildcard as the scheme:
```
Fatal Error:- Environment variable HASURA_GRAPHQL_CORS_DOMAIN: invalid domain: '*://localhost'. All domains should have scheme + (optional wildcard) host + (optional port)
```

Right now my solution is to allow any domain for CORS... I know I could probably setup a reverse proxy to handle CORS, but I feel like this could be easily supported using custom schemes, or wildcard as the scheme in Hasura directly.

Let me know if I can be of any help to support this! Thank you!

### PR DESCRIPTION
**Pull request by lilred**
### Description

Enables the use of Hasura server with custom URI schemes, necessary for e.g. Capacitor.

### Changelog

- [x] `CHANGELOG.md` is updated with user-facing content relevant to this PR. If no changelog is required, then add the `no-changelog-required` label.

### Affected components

- [x] Server

### Related Issues

#5818 

### Solution and Design

Accept any non-empty string not containing a colon as a scheme name.

### Steps to test and verify

Configure CORS config with an entry starting in e.g. `capacitor://`.

### Limitations, known bugs & workarounds

This probably gives an atrocious error message if the user forgets the colon in the scheme separator.

#### Catalog upgrade

Does this PR change Hasura Catalog version?
- [x] No
- [ ] Yes
  - [ ] Updated docs with SQL for downgrading the catalog 

#### Metadata

Does this PR add a new Metadata feature?
- [x] No
- [ ] Yes
  - Does `run_sql` auto manages the new metadata through schema diffing?
    - [ ] Yes
    - [x] Not required
  - Does `run_sql` auto manages the definitions of metadata on renaming?
    - [ ] Yes
    - [x] Not required
  - Does `export_metadata`/`replace_metadata` supports the new metadata added?
    - [ ] Yes
    - [x] Not required

#### GraphQL
- [x] No new GraphQL schema is generated
- [ ] New GraphQL schema is being generated:
   - [ ] New types and typenames are correlated
   
   

#### Breaking changes

- [x] No Breaking changes
- [ ] There are breaking changes:

  1. Metadata API

     Existing `query` types:
     - [ ] Modify `args` payload which is not backward compatible
     - [ ] Behavioural change of the API
     - [ ] Change in response `JSON` schema
     - [ ] Change in error code
     

  2. GraphQL API

     Schema Generation:
     
     
     - [ ] Change in any `NamedType`
     - [ ] Change in table field names
     

     Schema Resolve:-
     
     - [ ] Change in treatment of `null` value for any input fields 
     

  3. Logging

     - [ ] Log `JSON` schema has changed
     - [ ] Log `type` names have changed

### CODE REVIEW

**Review by GavinRay97 — APPROVED** [review]

**Comment by GavinRay97** [comment]
Note from conversation with Auke:

Do we want to modify this to restrict it to only spec-compliant URL schemes?
From a security standpoint, if it's a non-issue, maybe not. But for reference, here's the standards for schemes:

https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Definition

> The URI comprises:

> A non-empty scheme component followed by a colon (:), consisting of a sequence of characters beginning with a letter and followed by any combination of letters, digits, plus (+), period (.), or hyphen (-). Although schemes are case-insensitive, the canonical form is lowercase and documents that specify schemes must do so with lowercase letters. Examples of popular schemes include http, https, ftp, mailto, file, data, and irc. URI schemes should be registered with the Internet Assigned Numbers Authority (IANA), although non-registered schemes are used in practice.[b]

Tl;dr URL scheme =  any combination of alphanumeric or `+`/`.`/`-` characters

**Comment by lilred** [comment]
I'd be happy to make this change if required, however I would need to set up Hasura server development on my machine. I saw [CONTRIBUTING.md](https://github.com/hasura/graphql-engine/blob/master/server/CONTRIBUTING.md), which is helpful. Do you recommend any text editor/plugin combination?

**Comment by abooij** [comment]
@lilred Yes, that guide is certainly valuable. In particular, I'd advise you to follow the suggestion to use ghcup to install the correct GHC+cabal versions. And do follow the instructions precisely. In particular it's important to compile the console correctly if you want to test your work. (The Haskell side can be built independently, however a full development environment would also include a working console.)

Editors are, of course, a rather personal choice. Many Haskell developers use emacs. But I'd only recommend that for this project if you are already familiar with that - it's a big beast to learn to use. Visual Studio Code also has a number of packages that make it into an excellent Haskell editor, and offers a perhaps more welcoming experience.

**Comment by mehtavishwa30** [comment]
Hello @lilred!

Thanks so much for your contribution to Hasura OSS during the Hacktoberfest 2020! We'd love to send you some Hasura swag as a token of appreciation and you will be getting an email with the swag form shortly. ✨

**Comment by sdahlbac** [comment]
what is needed to get this one forward?

**Comment by lilred** [comment]
I haven't yet found the time to do this, things have gotten crazy at work. It's still on my personal backlog but consider it up for grabs.

**Comment by lilred** [comment]
The CI is behaving weirdly, which may be because this is an old PR. I'll close and reopen it.
